src/software/logger/proto_logger.cpp

```diff
@@ -0,0 +1,192 @@
+#include "proto_logger.h"
+
+#include <google/protobuf/message.h>
+#include <zlib.h>
+
+#include <chrono>
+#include <ctime>
+#include <experimental/filesystem>
+#include <fstream>
+#include <iomanip>
+#include <optional>
+#include <vector>
+
+#include "base64.h"
+#include "shared/constants.h"
+
+ProtoLogger::ProtoLogger(const std::string& log_path,
+                         std::function<double()> time_provider,
+                         const bool friendly_colour_yellow)
+    : log_path_(log_path),
+      time_provider_(time_provider),
+      friendly_colour_yellow_(friendly_colour_yellow),
+      stop_logging_(false),
+      buffer_(PROTOBUF_BUFFER_SIZE, true)
+{
+    start_time_ = time_provider_();
+
+    // Create a folder for the logs with the formatted current time
+    std::time_t t = std::time(nullptr);
+    std::tm tm    = *std::localtime(&t);
+    std::stringstream ss;
+    ss << std::put_time(&tm, REPLAY_FILE_TIME_FORMAT.data());
+    log_folder_ = log_path_ + "/" + REPLAY_FILE_PREFIX + ss.str() + "/";
+    std::experimental::filesystem::create_directories(log_folder_);
+
+    // Start logging in a separate thread
+    log_thread_ = std::thread(&ProtoLogger::logProtobufs, this);
+}
+
+ProtoLogger::~ProtoLogger()
+{
+    flushAndStopLogging();
+}
+
+void ProtoLogger::saveSerializedProto(const std::string& protobuf_type_full_name,
+                                      const std::string& serialized_proto)
+{
+    buffer_.push({
+        .protobuf_type_full_name = protobuf_type_full_name,
+        .serialized_proto        = serialized_proto,
+        .receive_time_sec        = time_provider_() - start_time_,
+    });
+}
+
+void ProtoLogger::logProtobufs()
+{
+    unsigned int replay_index = 0;
+
+    try
+    {
+        while (!shouldStopLogging())
+        {
+            std::string log_file_path =
+                log_folder_ + std::to_string(replay_index) + "." + REPLAY_FILE_EXTENSION;
+
+            gzFile gz_file = gzopen(log_file_path.c_str(), "wb");
+            if (!gz_file)
+            {
+                std::string error_msg =
+                    "Failed to open gzip file. Error: " + std::string(strerror(errno));
+                throw std::runtime_error(error_msg);
+            }
```

Comment:
From the look of things, this is the only line that throws an exception. 

If this is true I think it would be better to just log and break out of the loop instead of having a try catch statement?

Reply:
Good point. I've updated it to just return for now. If this does become a common error, we could look into better alternative ways of handling it.